src/queue/priority.rs

```diff
@@ -0,0 +1,492 @@
+/// readonly acess to a popped Entry.
+struct Slot<T> {
+    cell: AtomicCell<Option<T>>,
+}
+
+impl<T> Slot<T> {
+    fn new(value: T) -> Self {
+        Self {
+            cell: AtomicCell::new(Some(value)),
+        }
+    }
+
+    fn take(&self) -> Option<T> {
+        self.cell.take()
+    }
+}
+
+/// The runner for priority queues.
+///
+/// The runner helps collect additional information to support auto-adjust task level.
+/// [`PriorityRunnerBuiler`] is the [`RunnerBuilder`] for this runner.
+pub struct PriorityRunner<R> {
+    inner: R,
+    local_level0_elapsed_us: LocalIntCounter,
+    local_total_elapsed_us: LocalIntCounter,
+}
+
+impl<R, T> Runner for PriorityRunner<R>
+where
+    R: Runner<TaskCell = T>,
+    T: TaskCell,
+{
+    type TaskCell = T;
+
+    fn start(&mut self, local: &mut Local<T>) {
+        self.inner.start(local)
+    }
+
+    fn handle(&mut self, local: &mut Local<T>, mut task_cell: T) -> bool {
+        let extras = task_cell.mut_extras();
+        let level = extras.current_level();
+        let total_running_time = extras.total_running_time.clone();
+        let begin = Instant::now();
+        let res = self.inner.handle(local, task_cell);
+        let elapsed = begin.elapsed();
+        let elapsed_us = elapsed.as_micros() as u64;
+        if let Some(ref running_time) = total_running_time {
+            running_time.inc_by(elapsed);
+        }
+        if level == 0 {
+            self.local_level0_elapsed_us.inc_by(elapsed_us);
+        }
+        self.local_total_elapsed_us.inc_by(elapsed_us);
+        let local_total = self.local_total_elapsed_us.get();
+        if local_total > FLUSH_LOCAL_THRESHOLD_US {
+            self.local_level0_elapsed_us.flush();
+            self.local_total_elapsed_us.flush();
+        }
+        res
+    }
+
+    fn pause(&mut self, local: &mut Local<Self::TaskCell>) -> bool {
+        self.inner.pause(local)
+    }
+
+    fn resume(&mut self, local: &mut Local<Self::TaskCell>) {
+        self.inner.resume(local)
+    }
+
+    fn end(&mut self, local: &mut Local<Self::TaskCell>) {
+        self.inner.end(local)
+    }
+}
+
+/// The runner builder for priority task queues.
+///
+/// It can be created by [`Builder::runner_builder`].
+pub struct PriorityRunnerBuiler<B> {
+    inner: B,
+    level0_elapsed_us: IntCounter,
+    total_elapsed_us: IntCounter,
+}
+
+impl<B, R, T> RunnerBuilder for PriorityRunnerBuiler<B>
+where
+    B: RunnerBuilder<Runner = R>,
+    R: Runner<TaskCell = T>,
+    T: TaskCell,
+{
+    type Runner = PriorityRunner<R>;
+
+    fn build(&mut self) -> Self::Runner {
+        PriorityRunner {
+            inner: self.inner.build(),
+            local_level0_elapsed_us: self.level0_elapsed_us.local(),
+            local_total_elapsed_us: self.total_elapsed_us.local(),
+        }
+    }
+}
+
+/// The configurations of priority task queues.
+pub struct Config {
+    name: Option<String>,
+    level_time_threshold: [Duration; LEVEL_NUM - 1],
```

Comment:
can't configure this field?

Reply:
It's the same as multi-level, not configurable currently